notation/notation.go

```diff
@@ -0,0 +1,295 @@
+	ociDesc, err := repo.Resolve(ctx, reference)
+	if err != nil {
+		return Descriptor{}, err
+	}
+	desc := notationDescriptorFromOCI(ociDesc)
+	sig, signerInfo, err := signer.Sign(ctx, desc, envelopeMediaType, opts)
+	if err != nil {
+		return Descriptor{}, err
+	}
+	annotations, err := generateAnnotations(signerInfo)
+	if err != nil {
+		return Descriptor{}, err
+	}
+	_, _, err = repo.PushSignature(ctx, sig, envelopeMediaType, ociDesc, annotations)
+	if err != nil {
+		return Descriptor{}, err
+	}
+
+	return desc, nil
+}
+
+// VerifyOptions contains parameters for Verifier.Verify.
+type VerifyOptions struct {
+	ArtifactReference string
+	// SignatureMediaType is the envelope type of the signature.
+	// Currently both `application/jose+json` and `application/cose` are
+	// supported.
+	SignatureMediaType string
+	PluginConfig       map[string]string
+}
+
+// VerificationResult encapsulates the verification result (passed or failed)
+// for a verification type, including the desired verification action as
+//
+//	specified in the trust policy
+type VerificationResult struct {
+	// Success is set to true if the verification was successful
+	Success bool
+	// Type of verification that is performed
+	Type trustpolicy.ValidationType
+	// Action is the intended action for the given verification type as defined
+	// in the trust policy
+	Action trustpolicy.ValidationAction
+	// Err is set if there are any errors during the verification process
+	Error error
+}
+
+// VerificationOutcome encapsulates the SignerInfo (that includes the details of
+// the digital signature)
+// and results for each verification type that was performed
+type VerificationOutcome struct {
+	SignatureBlobDescriptor *ocispec.Descriptor
+	// EnvelopeContent contains the details of the digital signature and
+	// associated metadata
+	EnvelopeContent *signature.EnvelopeContent
+	// VerificationLevel describes what verification level was used for
+	// performing signature verification
+	VerificationLevel *trustpolicy.VerificationLevel
+	// VerificationResults contains the verifications performed on the signature
+	// and their results
+	VerificationResults []*VerificationResult
+	// SignedAnnotations contains arbitrary metadata relating to the target
+	// artifact that was signed
+	SignedAnnotations map[string]string
+	// Error that caused the verification to fail (if it fails)
+	Error error
+}
+
+// Verifier is a generic interface for verifying an artifact.
+type Verifier interface {
+	// Verify verifies the signature blob and returns the verified descriptor
+	// upon successful verification.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions, outcome *VerificationOutcome) (Descriptor, error)
+
+	// TrustPolicyDocument gets the validated trust policy document.
+	TrustPolicyDocument() (*trustpolicy.Document, error)
```

Comment:
This is because Verifier is an interface, it does not contain field trust policy. TrustPolicyDocument() is needed in notation.Verify function, as it takes the Verifier interface as input instead of an implementation.